File: modules/crypto-core/src/main/java/com/shterneregen/securelan/crypto/model/HybridEncryptedData.java
package com.shterneregen.securelan.crypto.model;

import java.util.Arrays;
import java.util.Objects;

public record HybridEncryptedData(byte[] encryptedKey, byte[] iv, byte[] cipherText) {
    public HybridEncryptedData {
        Objects.requireNonNull(encryptedKey, "encryptedKey");
        Objects.requireNonNull(iv, "iv");
        Objects.requireNonNull(cipherText, "cipherText");
        encryptedKey = encryptedKey.clone();
        iv = iv.clone();
        cipherText = cipherText.clone();
    }

    @Override
    public byte[] encryptedKey() {
        return encryptedKey.clone();
    }

    @Override
    public byte[] iv() {
        return iv.clone();
    }

    @Override
    public byte[] cipherText() {
        return cipherText.clone();
    }

    @Override
    public String toString() {
        return "HybridEncryptedData[encryptedKey=" + encryptedKey.length
                + " bytes, iv=" + iv.length
                + " bytes, cipherText=" + cipherText.length + " bytes]";
    }
}

File: modules/crypto-core/src/main/kotlin/com/shterneregen/securelan/crypto/model/HybridEncryptedData.kt
package com.shterneregen.securelan.crypto.model

import java.util.Objects

class HybridEncryptedData(
    encryptedKey: ByteArray,
    iv: ByteArray,
    cipherText: ByteArray,
) {
    private val encryptedKey: ByteArray = Objects.requireNonNull(encryptedKey, "encryptedKey").clone()
    private val iv: ByteArray = Objects.requireNonNull(iv, "iv").clone()
    private val cipherText: ByteArray = Objects.requireNonNull(cipherText, "cipherText").clone()

    fun encryptedKey(): ByteArray = encryptedKey.clone()

    fun iv(): ByteArray = iv.clone()

    fun cipherText(): ByteArray = cipherText.clone()

    override fun equals(other: Any?): Boolean =
        this === other ||
            (other is HybridEncryptedData &&
                encryptedKey == other.encryptedKey &&
                iv == other.iv &&
                cipherText == other.cipherText)

    override fun hashCode(): Int = Objects.hash(encryptedKey, iv, cipherText)

    override fun toString(): String =
        "HybridEncryptedData[encryptedKey=${encryptedKey.size} bytes, iv=${iv.size} bytes, cipherText=${cipherText.size} bytes]"
}
